### ISSUE TITLE
`dead_code` lint pass never considers inherent associated types alive

### ISSUE THREAD

**Issue by fmease** [rust-lang/rust#110332]
The following code triggers `dead_code` even though it should not.

```rust
#![feature(inherent_associated_types)]
#![allow(incomplete_features)]

fn main() {
    let _: Struct::Item = ();
}

struct Struct;
impl Struct { type Item = (); }
```

```
warning: associated type `Item` is never used
 --> min.rs:9:20
  |
9 | impl Struct { type Item = (); }
  | -----------        ^^^^
  | |
  | associated type in this implementation
  |
  = note: `#[warn(dead_code)]` on by default

warning: 1 warning emitted
```

Actually until recently, `dead_code` never triggered at all for inherent associated types.
PR #110277 was just merged which started visiting more associated items than before
exposing this bug in the `dead_code` pass.
The pass never considers IATs as *live symbols* (`live_symbols`).

This feels like a regression but to be technically correct it is not, we switched from one
extreme to the other. Thus not marking as such.

@rustbot label A-lint F-inherent_associated_types

**Comment by Ezrashaw** [comment]
@fmease

Implementation is blocked on the following issue (not sure I'll be able to fix this):
Inherent associated types aren't resolving paths properly. This can be observed from the code example provided in the OP with `-Zunpretty=hir-tree`:
```
PathSegment {
    ident: Item#0,
    hir_id: HirId(DefId(0:3 ~ rust_out[dfb5]::main).5),
    res: Err,
    args: None,
    infer_args: false,
}
```
(note the `res: Err`).

I believe my impl of this issue is blocked on this.

**Comment by fmease** [comment]
Right, I stopped investigating after noticing `Res::Err`. I believe we are not writing back the resolution / typeck result for IATs. Gonna check if I can fix the latter with or without `AliasKind::Inherent` (#109410).

**Comment by Ezrashaw** [comment]
Ok, I have a pretty trivial fix ready for when that PR is merged.

**Comment by fmease** [comment]
I might be slightly wrong with my analysis but I think this roughly sums it. I'm gonna drop a message on Zulip later though to double-check.

As a *mitigation*, we can consider IATs always alive, too.

**Comment by Ezrashaw** [comment]
@fmease IIUC (and I don't know a lot about the compiler), it should be enough *for now* to only have the paths in the function body. We can then resolve it to a `DefId` and mark the inherent associated type live.

For the case where the type isn't referenced from within a function, eg the following:
```rust
fn main() {
    f::<Struct>();
}

struct Struct;
impl Struct { type Item = usize; }

struct Foo { s: Struct::Item }
```
we can leave it as is for now, until a better solution is found.

As an aside (and this might be naive), we should never have `Res::Err`, right? If we truly can't resolve a path, then we have to emit a error, else there is a correct resolution?

**Comment by fmease** [comment]
Sorry for the delay. I'm gonna look into the matter one more time to see what we can do about paths outside of `FnCtxt`s.

If I don't find anything useful, I will open a PR to update the `Res` of paths in `FnCtxt`s so you can swoop in and solve half the issue. For that, personally I'd prefer to consider the assoc tys as always alive if we cannot tell for sure.

The presence of `Res::Err` in legitimate paths *after* `hir_typeck` has run defo sounds fishy (both for IATs & for the stable example I gave above). Afaik, type-dependent paths are `Res::Err` after `rustc_resolve` has run and they are then updated on a case by case basis during typeck.

**Comment by Ezrashaw** [comment]
All good, time as much time as you need. I agree we should consider `Res::Err` IATs live if their isn't a good fix, however that should probably be looked at again before IATs are stabilized.

**Comment by fmease** [comment]
BoxyUwu is already working on this specific issue as well on as larger architectural blockers.
